### ISSUE TITLE
[flutter release] race condition when uploading release artifact

### ISSUE THREAD

**Issue by XilaiZhang** [flutter/flutter#142325]
On release website https://docs.flutter.dev/release/archive?tab=macos, sometimes only the x64 version of artifact is available, and the arm64 version release artifact is missing.

context: in https://chat.google.com/room/AAAAc_4rqiI/cI_UlyEqeqc @godofredoc provided insights and pointed out that this might be caused by a race condition when uploading the artifacts.

**Comment by godofredoc** [comment]
This is happening because of a race condition when writing to gs://flutter_infra_release/releases/releases_macos.json.

The two packaging builds run in parallel which causes an execution similar to:

* Mac_arm64 packaging build reads the json file from gcs
* Mac_x64 packaging builds reads the json file from gcs
* Each build indepedently modifies the file.
* Mac_arm64 finishes first and uploads the file.
* Mac_x64 finishes uploading his version of the file overwriting the arm64 entry.

This can be fixed using a lock file to ensure atomic read/writes to the file.

**Comment by godofredoc** [comment]
@CaseyHillers given this has already happened a couple of times and may also happen on stable releases we may want to consider this issue high priority.

**Comment by Jasguerrero** [comment]
Duplicate of https://github.com/flutter/flutter/issues/135328
